### ISSUE TITLE
[Feature] migrate to Typescript + support puppeter 8.0.0 (writen in Typescript)

### ISSUE THREAD

**Issue by UrielCh** [berstend/puppeteer-extra#465]
**Feature request**

Hi, now that *Puppeteer* moved to Typescript, and *Playwright* is already written in typescript.

It's time to switch to it.

I open 2 PR about that:
- #455 update code to support pptr 2..8
- #461 convert all code to typescript

Now I would like to improve code using more advanced Typescript features, for that I need to change all exports to modern ones.

Can you give me some feedback?

In my next PR, all plugin will enforce their options typing.
and the activation of each plugin will update puppeteer types, to allow IntelliSense to suggest new methods added by plugins.
and then I will be able to plug my plugin(s) to puppeteer-extra this project.

**Comment by berstend** [comment]
First of all: Thanks for showing the initiative and aiming at supporting the project :-)

It would've been good though to open an issue like this before investing any work, so we have a chance to discuss and coordinate these changes. :-)

The project is on the verge of switching to a major rewrite in the `automation-extra` branch (#303), which features improved type safety and will make the current master branch that you've worked against obsolete.

As you've probably noticed the core of this project is written in TS already (the plugin framework, the base plugin class as well as a few plugins themselves), so we're clearly interested in a smooth experience for TS users.

A couple of things I'd like to discuss and hear your opinion on, before you put more work in PRs:

**Backwards compatibility**
- The project aims to support older puppeteer versions as long as it reasonably can, this means `2.1.1` as a minimum
- Looking at #455 your changes seem to require pptr 8 (`Puppeteer.BrowserLaunchArgumentOptions`)

We need to find a way to support typings both for pptr 2/5 (which depend on `@types/puppeteer`) as well as the newer version with their partially broken built-in typings.

You didn't really address that concern I raised in #455, or at least not in a way I noticed it. :-)

#461 is a massive PR and it's hard for me to get the gist of it. The `configure.js` stuff looks interesting but I need more context of what this is doing and how it helps with the aforementioned backwards compatibility issue.

Do you have any opinion on how to best use types internally while allowing the end-user to use a different typings version? We currently hard-code `{playwright,puppeteer}-core` as a dependency as we internally need types and want to give non-TS users Intellisense out of the box - it could make sense to revisit `peerDependencies` for this though.

---

Overall:
- Breaking backwards compat and requiring pptr > 8 is **a massive change** and I only consider this as a last resort if we exhausted all other options. If it's really necessary this could be part of the #303 release, which will trigger a major version bump as well (readmes need to be updated accordingly)
- I can't realistically review these massive PRs and merge them into `automation-extra`, so we need to a) coordinate this more (hence this long text) or b) I use your PRs as reference and cherry-pick things myself when the time has come.

One other thing:
- I'm not 100% sure if I want to rewrite the stealth plugin in Typescript, as it makes the code less approachable and beginner friendly. I do however would like to see improved type safety, especially around option parameters. I was wondering if creating custom .d.ts files while retaining JS code would give us the best of both worlds. (Side-note: I'm currently prototyping a new `@extra/stealth` plugin which allows to override evasion plugin settings without having to remove/re-add the evasion plugin, that's written in TS and features full type-safety).

Thanks!

**Comment by UrielCh** [comment]
Hi, I saw the `automation-extra` branch, I will re-write my changes on the correct branch, this time it won't take long.
I mostly port your code to Typescript to read it, and to understand it.

For Backwards compatibility, I made a [script](https://github.com/UrielCh/puppeteer-extra/blob/ts/configure.js) to keep the v2.1.1 retro compatibility test working (but you already find it).

The only Backwards compatibility issue is the loss of the support of node v8.

I'm only working in Typescript for years, I converted a lot of codebases from javascript to Typescript, 96% of projects I converted were containing errors, that I discovered during the conversion. I only found a single project without errors, the guy is a genius, an autistic, or ... a machine. I have to get in touch with him/it(:robot:).

You can have a look at [puppeteer-jquery](https://www.npmjs.com/package/puppeteer-jquery), I will integrate it to puppeteer-extra, at the end.

I mostly convert code to typescript to let IntelliSense works instead of me :relaxed:
Have a look at my IntelliSense-oriented project [proxmox-api](https://www.npmjs.com/package/proxmox-api), I put an animated gif to show the auto-completion.

A Typescript enabled evasion plugin, will allow the new feature to be displayed by IntelliSense.

Tell me when your `automation-extra` branch will be ready to get ported to Typescript and do not spend too much time on writing the d.ts files, I'm an expert in TS conversion.

Best regards, Uriel.

**Comment by UrielCh** [comment]
I have just ported my [puppeteer-jquery](https://www.npmjs.com/package/puppeteer-jquery) to [playwright-jquery](https://www.npmjs.com/package/playwright-jquery).

**Comment by UrielCh** [comment]
Hi,

Can you please merge changes from the `master` branch to `automation-extra` branch?

I would like to play with *Puppeteer* / *Playwright* soon.

I only need a few days to make a new PR from `automation-extra` branch (or multiple smaller ones).

**Comment by berstend** [comment]
This should now be fixed through #555:
```
Successfully published:
 - puppeteer-extra-plugin-adblocker@2.12.0
 - puppeteer-extra-plugin-anonymize-ua@2.3.0
 - puppeteer-extra-plugin-block-resources@2.3.0
 - puppeteer-extra-plugin-click-and-wait@2.3.0
 - puppeteer-extra-plugin-devtools@2.3.0
 - puppeteer-extra-plugin-flash@2.3.0
 - puppeteer-extra-plugin-font-size@2.3.0
 - puppeteer-extra-plugin-recaptcha@3.5.0
 - puppeteer-extra-plugin-repl@2.3.0
 - puppeteer-extra-plugin-stealth@2.8.0
 - puppeteer-extra-plugin-user-data-dir@2.3.0
 - puppeteer-extra-plugin-user-preferences@2.3.0
 - puppeteer-extra-plugin@3.2.0
 - puppeteer-extra@3.2.0
```

As mentioned previously I can't merge your PRs as is @UrielCh but they've been a helpful reference (and there's more future TS improvements to come)
